contents/blog/clickhouse-materialized-columns.md

```diff
@@ -0,0 +1,148 @@
+---
+date: 2021-09-09
+title: Speeding up ClickHouse with Materialized Columns
+rootPage: /blog
+sidebar: Blog
+showTitle: true
+hideAnchor: true
+categories: engineering
+author: karl-aksel-puulmann
+featuredImage: ../images/blog/automating-software-company-github-actions.png
+featuredImageType: full
+---
+
+Did you know ClickHouse supports speeding up queries up to an order of magnitude by using Materialized Columns? This rarely-used feature can be used to create new columns on the fly from existing data, speeding up queries.
+
+In this post, I’ll walk through an example query optimization in which materialized columns are well suited.
+
+Consider the following schema:
+
+```sql
+CREATE TABLE events (
+    uuid UUID,
+    event VARCHAR,
+    timestamp DateTime64(6, 'UTC'),
+    properties_json VARCHAR,
+)
+ENGINE = MergeTree()
+ORDER BY (toDate(timestamp), event, uuid)
+PARTITION BY toYYYYMM(timestamp)
+```
+
+Each event has an ID,  event type, timestamp, and a JSON representation of event properties. The properties can include the current URL and any other user-defined properties that describe the event (e.g. NPS survey results, person properties, timing data, etc).
+
+This table can be used to store a lot of analytics data and is similar to what we use at PostHog.
+
+If we wanted to query login page pageviews in August, the query would look like this:
+
+```sql
+SELECT count(*)
+FROM events
+WHERE event = '$pageview'
+  AND JSONExtractString(properties_json, '$current_url') = 'https://app.posthog.com/login'
+  AND timestamp >= '2021-08-01'
+  AND timestamp < '2021-09-01'
+```
+
+On a large test dataset this query takes a while complete, while without the url filter the query is almost instant. Adding even more filters just makes the query slower and slower. Let’s dig in why!
+
+## Looking at flamegraphs
+
+ClickHouse has great tools for introspecting queries. Looking at `system.query_log`  we can see that the query:
+
+- Took 3433ms
+- Read 79.17 GiB from disk
+
+To dig even deeper, we can use clickhouse-flamegraph to peek into what the CPU is doing during query execution.
+
+[![Flamegraph](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)
```

Comment:
Unfortunately this link doesn't work because images are rendered in a particular way at Gatsby build, which works for image tags (the link gets changed to the "compiled" version of the image in it), but doesn't apply to links. (@smallbrownbike?)

BTW could maybe add a caption that this SVG is interactive, because it's pretty cool.

Reply:
Apart from the caption, is there any way we can link to a working rendition of the SVG?

Reply:
Made a hacky solution after spending way way too much time on this. SVG import broke inline js for this so that's out. :crossed_fingers: